File: exercise_three/scripts/add_client.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import rospy
from exercise_three.srv import *

def client():
    # 初始化节点
    rospy.init_node('add_client')
    rospy.loginfo('add_client node is Ready!')
    # 在target_detection服务的服务端启动前,服务的调用一直处于阻塞状态
    rospy.wait_for_service("call_add")
    # 创建对象client用来调用target_detection服务
    client = rospy.ServiceProxy('call_add', CallAdd)
    x = rospy.get_param('~x', 1)
    y = rospy.get_param('~y', 2)
    z = rospy.get_param('~z', 3)

    try:
        req = CallAddRequest()
        req.addend.x = x
        req.addend.y = y
        req.addend.z = z
        rospy.loginfo("parameter:x=%f,y=%f,z=%f",req.addend.x,req.addend.y,req.addend.z);

        # 进行服务调用,res保存服务端返回的应答数据
        res = client(req)
        rospy.loginfo("sum: %f",res.sum);

    except rospy.ServiceException as e:
        rospy.logerr("Service call failed: %s"%e)

if __name__ == '__main__':
    client()

File: exercise_three/src/add_client.cpp
#include <ros/ros.h>
#include "exercise_three/CallAdd.h"

int main(int argc, char** argv){
  // 初始化ROS节点
  ros::init(argc, argv, "add_client");	

  ROS_INFO("add_client node is Ready!");
  // 创建节点句柄
  ros::NodeHandle nh;
  // add_server服务的服务端启动前，服务的调用一直处于阻塞状态
  ros::service::waitForService("call_add");

  // 创建一个client，请求add_three service, service消息类型是 exercise_three::Calladd
  ros::ServiceClient client = nh.serviceClient<exercise_three::CallAdd>("call_add");

  float x,y,z;
  ros::param::param<float>("~x", x, 1);
  ros::param::param<float>("~y", y, 2);
  ros::param::param<float>("~z", z, 3);


  // 创建exercise_three::Calladd类型的service消息
  exercise_three::CallAdd srv;
  srv.request.addend.x = x;
  srv.request.addend.y = y;
  srv.request.addend.z = z;
  
  ROS_INFO("parameter:x=%f,y=%f,z=%f",srv.request.addend.x,srv.request.addend.y,srv.request.addend.z);

  // 发布service请求，等待加法运算的应答结果
  if (client.call(srv)){
    ROS_INFO("Sum: %f", srv.response.sum);
  }else{
    ROS_ERROR("Failed to call service add_server");
    return 1;
  }
	
  return 0;
};


